Adaptive lasso

henridwyer opened the pull request:
Implement adaptive Lasso, and resolves issue #555

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scikit-learn/scikit-learn/4912)

<!-- Reviewable:end -->

henridwyer committed:
Implemented adaptive lasso

henridwyer committed:
Add a test for the adaptive lasso

henridwyer committed:
Added example for adaptive lasso

henridwyer committed:
Update title and introduction to include adaptive lasso

henridwyer committed:
Fixed adaptive lasso import

henridwyer committed:
fix travis errors

avoid mutating eps
change array division to use np.divide

henridwyer committed:
remove lambda function

amueller commented:
Cool. Can you maybe post the plot from the example? Have you done any real-world comparison / benchmarks?

amueller commented:
hm, your references are much older than the nips reference. What is the difference in the nips paper?

henridwyer commented:
I couldn't find the nips paper, so I am not sure. Do you have a link to it?

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2034,138 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+        1/n * ||y - X Beta||^2_2 + alpha * w |Beta|_1
+
+    Where w is a weight vector calculated in the previous stage by::
+        w_j = alpha/(|Beta_j|^gamma + eps)
```
you should clarify what cost function you're actually minimizing with this reweighting scheme.

see this old snippet of mine:

https://gist.github.com/agramfort/1610922

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2034,138 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+    
+    def fit(self, X, y):
+        """
+        Fit Lasso models, each with coordinate descent.
```
this doc is not properly formatted

see pep 257

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2034,138 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+        Fit Lasso models, each with coordinate descent.
+        Parameters
+        -----------
+        X : ndarray or scipy.sparse matrix, (n_samples, n_features)
```
(n_samples, n_features) -> shape (n_samples, n_features)

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2034,138 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+        -----------
+        X : ndarray or scipy.sparse matrix, (n_samples, n_features)
+            Data
+        y : ndarray, shape = (n_samples,) or (n_samples, n_targets)
```
no equal after shape

agramfort commented:
can you please share a screenshot of the output of the example?

Commit:
updated code and docstring for pep8 and pep 257

henridwyer commented:
Here is the example graph:
![adaptive lasso example graph](https://cloud.githubusercontent.com/assets/3540985/8513168/0622a52c-232e-11e5-8871-c0ee85ce2507.png)

I clarified the objective in the docstring.

I updated the docstring to be compliant with pep257 (I think?). For the fit method, I had copied the docstring from the ElasticNet fit function so I updated that one too.

I also changed the references a bit.

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2034,168 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+    The Adaptive Lasso and Its Oracle Properties
+    Journal of the American Statistical Association
+    """
+    def __init__(self, n_lasso_iterations=2, gamma=1, alpha=1.0,
```
n_lasso_iterations=2 seems small. Also is eps=1e-3 a good default? I personally don't use eps (set it to zero) and I discard features once they have been zeroed.

henridwyer replied:
Well I was thinking that the adaptive lasso corresponds to 2 steps, but this could be increased to more (5?).

About eps: 0.001 was the value they used in another article (added ref), that's why I figured it was a sensible default.

agramfort replied:
>
> Well I was thinking that the adaptive lasso corresponds to 2 steps, but
> this could be increased to more (5?).
>
you should iterate until the cost function stop decreasing up to a given
tolerance. My experience with neuroscience data is that 5 to 10 iterations
is enough.

> About eps: 0.001 was the value they used in another article (added ref),
> that's why I figured it was a sensible default.
>
hum ok...

agramfort commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -604,3 +604,12 @@ def test_sparse_dense_descent_paths():
         _, coefs, _ = path(X, y, fit_intercept=False)
         _, sparse_coefs, _ = path(csr, y, fit_intercept=False)
         assert_array_almost_equal(coefs, sparse_coefs)
+
```
2 lines between functions

Commit:
added eps ref, and updated docstring with blank lines

Commit:
added line in test

Commit:
added iterations until convergence of objective

henridwyer commented:
I added calculation of the objective function, and iterations until the progress of the objective is below ada_tol

henridwyer commented:
I changed n_lasso_iterations to max_lasso_iterations and set the default value to 20

Commit:
fixed loop

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2035,222 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+    Gasso, G., Rakotomamonjy, A., & Canu, S.
+    Recovering Sparse Signals With a Certain Family of Nonconvex
+    Penalties and DC Programming
+    IEEE Trans. Signal Process., 4686-4698.
```
what is the year?

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2035,222 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+
+    Zou, H., & Li, R.
+    One-step sparse estimates in nonconcave penalized likelihood models.
+    The Annals of Statistics Ann. Statist., 1509-1533.
```
year?

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2035,222 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+    Buhlmann, P., & Meier, L.
+    Discussion: One-step sparse estimates in nonconcave penalized likelihood
+    models.
+    The Annals of Statistics Ann. Statist., 1534-1541.
```
year?

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2035,222 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+
+    Zou, Hui.
+    The Adaptive Lasso and Its Oracle Properties
+    Journal of the American Statistical Association
```
year?

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2035,222 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+    The optimization objective for the AdaptiveLasso is::
+
+        (1 / (2 * n_samples)) * ||y - X Beta||^2_2
+        + alpha * p(|Beta|_1 + eps)
```
you should parametrize it with:

alpha * (|Beta|_1 + eps)^p

and then the solver to use p=0.5 by default. It's easier to think of the exponent of the reg than its derivative.

agramfort replied:
you should also check that p \in ] 0, 1 [ otherwise it does not work.

agramfort commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -604,3 +604,13 @@ def test_sparse_dense_descent_paths():
+    max_iter = 150
+    clf1 = AdaptiveLasso(max_iter=max_iter, max_lasso_iterations=1).fit(X, y)
+    clf2 = Lasso(max_iter=max_iter).fit(X, y)
+    assert_array_almost_equal(clf1.coef_,clf2.coef_)
```
can you test that adaptive lasso yields sparser estimates when exponent gets closer to 0?

amueller commented:
is this still work in progress or waiting for reviews?
I can't really judge how relevant / important this work is. @GaelVaroquaux @agramfort @mblondel might have more informed opinions?

agramfort commented:
most of my comments are sill not addressed apparently.

Adaptive Lasso works better in certain applications (eg. correlated designs). So I would be fine adding to sklearn given that it's not that many lines of code.

Commit:
added years to citations

henridwyer commented:
The reason I parametrized with the derivitive is to make it simpler to have p=ln, which would be more complicated without. What do you think?

agramfort commented:
fair enough for parametrization.

Please:
- address my comment on test
- run pep8 checker on your code

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2033,3 +2035,222 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+    The Adaptive Lasso and Its Oracle Properties
+    Journal of the American Statistical Association, 2006.
+    """
+    def __init__(self, max_lasso_iterations=20, gamma=1, alpha=1.0,
```
gamma=1 -> gamma=1.0

Commit:
added test for sparser coefficients

henridwyer commented:
I added a test for sparser coefficients - although maybe it could be more extensive - and removed the named lambda function (I think that was the pep8 issue)

Commit:
fix format for test

agramfort commented:
why did you close?

henridwyer commented:
I was planning on rebasing on master and then rewriting part of it to make it pass the tests, and then reopening. Is that a bad approach ?

agramfort commented:
Ok sounds good

Commit:
merge conflicts

Commit:
changed loop, added warm start

Commit:
added warm start

Commit:
refactor code

Commit:
pep8 and pep257

Commit:
fix pickling of penalty and loss functions

Commit:
changed loss function. fixed doc

Commit:
fix documentation

henridwyer commented:
I rewrote the class with a few major changes:
- changed the parameters. In particular there is now a penalty parameter to define which penalty to use.
- added the log and scad penalty as separate penalties. 
- calculated the loss at each iteration to check for convergence
- more complete, testing for sparser coefficients and testing that the loss decreases at each iteration, which it should according to the paper.

I had a bit of trouble with getting _p and _p_prime to work, since the pickling test would fail (originally I had a factory method). Is this the right way to implement those?

Commit:
numpy copyto -> copy

agramfort commented:
sleeping over it I feel this should be a metaestimator that allows you to reweight all linear models that expose a .coef_ parameter. It would then work for MultiTaskLasso, sparse logistic regression, Elastic-Net etc.

henridwyer commented:
so something like RANSACRegressor ?

agramfort commented:
yes but let's first maybe see what other core dev think?

we're quite slow adding new code in sklearn these days

henridwyer commented:
Tthinking about it a bit more, it makes sense to be able to use a multitask lasso and logistic regression - but I'm not sure about other estimators (I've never seen reweighted elasticnet).

henridwyer commented:
@agramfort any update?

danmackinlay commented:
Updated the reference request in the attached issue. FWIW, the key point here for me would be simply allowing a "weight" to each observation so that the user can choose their own reweighting scheme, Adaptive Lasso or one of the many others. This is the approach taken with R's `glmnet` and people regularly publish new papers testing out new and different weighting schemes.

jnothman commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2159,3 +2161,315 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+    The optimization objective for the AdaptiveLasso is::
+
+        (1 / (2 * n_samples)) * ||y - X Beta||^2_2
+        + alpha * \sum_j p(|Beta|_j)
```
I think that should be `|Beta_j|` not `|Beta|_j`

henridwyer commented:
@danmackinlay what API are you imagining? You would want more granularity than the penalty and q parameters ?

danmackinlay commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2159,3 +2161,315 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+
+        for k in xrange(1, self.max_lasso_iterations):
+            self.n_iter_ = k
+            weights = self._p_prime(np.abs(self.coef_))
```
This weighted fitting feels like it could potentially live in the parent class, since weights are not specific to the Adaptive Lasso but occur in other variant of the Lasso; the Adaptive Lasso is unique for how and when it chooses weights, not that it has weights.

danmackinlay commented:
@henridwyer sorry, that may not have been clear; For `AdaptiveLasso` I think the proposed API is great; rather it's the implementation as concerns the relationship between it and `Lasso`. If I had my brain engaged I would have been more clear. Trying again:

I believe weighted coefficient penalties are useful for many Lasso variants that could subclass `Lasso` in addition to `AdaptiveLasso` e.g. custom robust fitting, or a-priori-given weights because I don't want to penalise some coefficients etc. 
So my suggestion is that the weight calculation naturally belongs in Adaptive Lasso, but the weights should be optionally selectable in the parent `Lasso` class (as with e.g. the non-adaptive `penalty.weight` in [glmnet](https://cran.r-project.org/web/packages/glmnet/glmnet.pdf), and `AdaptiveLasso` could then pass those weights to `Lasso`

(I also think we should allow *observation* weights, like `glmnet`, FWIW, but that's a separate question.)

danmackinlay commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2159,3 +2161,315 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+        equivalent to a Lasso, solved by the :class:`Lasso`, and
+        max_lasso_iterations = 2 is equivalent to an adaptive Lasso.
+
+    penalty : 'scad' | 'log' | 'lq' (default='lq')
```
Inconsistency: in the actual constructor the default is `'log'`

danmackinlay commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2159,3 +2161,315 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+    The Adaptive Lasso and Its Oracle Properties
+    Journal of the American Statistical Association, 2006.
+    """
+    def __init__(self, max_lasso_iterations=30, penalty='log', q=None,
```
Given that this is called adaptive Lasso, the `log` penalty as default is might be surprising; it should probably default to a parameters from the original lasso paper, e.g. `penalty='lq', q=1`

agramfort commented:
it don't think it deserves to be a core estimator. It's simple to implement and can eventually be a simple example.

jodie-c commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2159,3 +2161,315 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+    ----------
+    coef_ : array, shape (n_features,) | (n_targets, n_features)
+        parameter vector (w in the cost function formula)
+
```
I think this should be `(Beta in the cost function formula)`

jodie-c commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -2159,3 +2161,315 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+                or self.max_lasso_iterations <= 0:
+            raise ValueError("Maximum number of Lasso iterations must be"
+                             " positive; got (max_iter=%r)" % self.max_iter)
+
```
ValueError raised for `max_iter` parameter value but conditioned on `max_lasso_iterations`
